EVG-2659 create config model

john-m-liu opened the pull request:
I didn't add extra functions to set parts of the model because I plan to have the API/UI update the entire document at once

john-m-liu commented on config.go:
```diff
@@ -26,46 +26,46 @@ var (
-	NumNewRepoRevisionsToFetch int
-	MaxRepoRevisionsToSearch   int
-	MaxConcurrentRequests      int
+	NumNewRepoRevisionsToFetch int `bson:"numnewreporevisionstofetch" json:"numnewreporevisionstofetch" yaml:"numnewreporevisionstofetch"`
```
done

john-m-liu commented on config.go:
```diff
@@ -81,126 +81,126 @@ type ClientConfig struct {
 // NotifyConfig hold logging and email settings for the notify package.
 type NotifyConfig struct {
-	SMTP *SMTPConfig `yaml:"smtp"`
+	SMTP *SMTPConfig `bson:"smtp" json:"smtp" yaml:"smtp"`
```
this one works as expected and is in the test

john-m-liu commented on config.go:
```diff
@@ -216,7 +216,7 @@ func (j JiraConfig) GetHostURL() string {
+	SMTP *SMTPConfig `bson:"smtp" json:"smtp" yaml:"smtp"`
 }
 type WriteConcern struct {
 	W        int    `yaml:"w"`
```
since this struct isn't going to be in the model at all, I'd prefer to leave it alone for now

bsamek commented on config.go:
```diff
@@ -26,46 +26,46 @@ var (
-	NumNewRepoRevisionsToFetch int
-	MaxRepoRevisionsToSearch   int
-	MaxConcurrentRequests      int
+	NumNewRepoRevisionsToFetch int `bson:"revs_to_fetch" json:"revisions_to_fetch" yaml:"numnewreporevisionstofetch"`
```
is the bson/json tag difference deliberate?

bsamek commented on config.go:
```diff
@@ -252,56 +252,55 @@ type LogBuffering struct {
-	PprofPort           string                    `yaml:"pprof_port"`
-	GithubPRCreatorOrg  string                    `yaml:"github_pr_creator_org"`
-	NewRelic            NewRelicConfig            `yaml:"new_relic"`
+	Database           DBSettings                `yaml:"database"`
```
Would you mind alphabetizing this list? It'll make is easier to read/understand diffs to it in the future?

john-m-liu commented:
I addressed the comments and also made a big refactor. I accidentally amended my old commit so you won't be able to diff correctly but everything is changed now. Currently this is my WIP, but I wanted to run this past everyone before I proceeded further since I'll just be adding each config section in the same way that has been added already

john-m-liu commented:
This is ready for cr now

bsamek commented on config.go:
```diff
@@ -22,51 +25,112 @@ var (
+	if err != nil && err.Error() == errNotFound {
+		return nil
+	}
+	return err
```
all getters and setters return err, should they wrap?

bsamek commented on config.go:
```diff
@@ -22,51 +25,112 @@ var (
+	})
+	return err
+}
+func (c *AuthConfig) validateAndDefault() error {
```
This function can error in several ways. It might make sense to use a catcher, to avoid the case where someone fixes one problem only to discover another (duplicate user, no github auth).

bsamek commented on config.go:
```diff
@@ -22,51 +25,112 @@ var (
+
+func (c *AuthConfig) id() string { return "auth" }
+func (c *AuthConfig) get() error {
+	err := legacyDB.FindOneQ(Collection, legacyDB.Query(byId(c.id())), c)
```
seems like legacyDB can just be db

bsamek commented on config.go:
```diff
@@ -319,16 +874,113 @@ func NewSettings(filename string) (*Settings, error) {
+		if !propVal.CanSet() {
+			catcher.Add(fmt.Errorf("unable to set field %s in %s", propName, sectionId))
+		}
+		propVal.Set(sectionVal)
```
this should be in an else

bsamek commented on config_registry.go:
```diff
@@ -0,0 +1,76 @@
+}
+
+type configSectionRegistry struct {
+	mux      *sync.RWMutex
```
nit: call it `mu` because `mux` is confusing for the router

bsamek commented on config_test.go:
```diff
@@ -4,9 +4,33 @@ import (
+)
+
+// TestConfig creates test settings from a test config.
+func testConfig() *Settings {
```
is there a test for the fallback default behavior?

john-m-liu replied:
added a test

tychoish commented on config_db.go:
```diff
@@ -0,0 +1,97 @@
+)
+
+var (
+	Collection  = "admin"
```
how needed is making this public? might need a better name?

john-m-liu replied:
It's used in other packages in tests. I renamed it

tychoish commented on config_test.go:
```diff
@@ -4,9 +4,33 @@ import (
+	}
+
+	return settings
+}
```
Is it reasonable to merge this with the testutils.TestConfig() method? (maybe my moving it there.)

john-m-liu replied:
Once this moves to a different package, the function will go away (it's just here for now to avoid an import cycle)

tychoish commented on config_test.go:
```diff
@@ -4,9 +4,33 @@ import (
+
+	if err = settings.Validate(); err != nil {
+		panic(err)
+	}
```
these panics (even in tests) seem less than awesome but maybe its alright?

john-m-liu replied:
This copy of the function will go away with the refactor, so I think it's find to keep here

tychoish commented on config_registry.go:
```diff
@@ -0,0 +1,76 @@
+	}
+	if catcher.HasErrors() {
+		panic(fmt.Sprintf("error registering config sections: %s", catcher.String()))
+	}
```
I might prefer having an init function in each of the file/files where a config block is defined that adds it to the registry, so that you can add a section editing only one or two files.

similarly, is it possible to write a test that will error if you add a section that isn't in the registry?

lastly, and along these lines, the registry is public, is there a way to move at least some of the code *out* of the top level package?

john-m-liu replied:
Discussed in person, I'll make these changes after refactoring these into a different package (when we get rid of the file entirely). The tests will already fail if you add a section that isn't in the registry because GetConfig() will error

tychoish commented on rest/data/interface.go:
```diff
@@ -143,14 +143,14 @@ type Connector interface {
-	GetAdminSettings() (*admin.AdminSettings, error)
-	SetAdminSettings(*admin.AdminSettings, *user.DBUser) error
+	GetAdminSettings() (*evergreen.Settings, error)
+	SetAdminSettings(*evergreen.Settings, *user.DBUser) error
```
the names of these methods seems potentially weird now.

john-m-liu replied:
renamed

tychoish commented on units/patch_intent_test.go:
```diff
@@ -11,7 +11,7 @@ import (
 	"github.com/evergreen-ci/evergreen/mock"
 	"github.com/evergreen-ci/evergreen/model"
-	"github.com/evergreen-ci/evergreen/model/admin"
+	
```
I'd avoid these blank lines in the import blocks

john-m-liu replied:
fixed

tychoish commented on config_registry.go:
```diff
@@ -0,0 +1,82 @@
+}
+
+type configSectionRegistry struct {
+	mu       *sync.RWMutex
```
I would not make this a pointer, so you can avoid needing to explicitly construct this value, because the registry.

tychoish commented on config_registry.go:
```diff
@@ -0,0 +1,82 @@
+	"github.com/mongodb/grip"
+)
+
+var registry *configSectionRegistry
```
can we name this configRegistry

(you can make more friendly accessor functions)
